File: CreatingAndCustomizingTheTouchBar/Objective-C/NSTouchBar Catalog/TestViewControllers/CandidateListViewController.m
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
View controller responsible for showing NSCandidateListTouchBarItem in an NSTouchBar instance for picking contacts as a use case.
*/

#import "CandidateListViewController.h"
@import Contacts;

static NSTouchBarCustomizationIdentifier CandidateCustomizationIdentifier = @"com.TouchBarCatalog.candidateListViewController";

@interface CandidateListViewController () <NSTouchBarDelegate, NSCandidateListTouchBarItemDelegate, NSTextFieldDelegate>

@property (strong) NSCandidateListTouchBarItem<CNContact *> *candidateListItem;
@property (weak) IBOutlet NSTextField *textField;

@end


#pragma mark -

@implementation CandidateListViewController

- (void)viewDidAppear
{
    [super viewDidAppear];
    
    // Use the candidate item instead of the NSTextField item.
    self.textField.automaticTextCompletionEnabled = NO;
    
    // Note: To show the NSTouchBar instance within this view controller, do this:
    // [self.view.window makeFirstResponder:self.view];
}

#pragma mark NSTouchBarProvider

- (NSTouchBar *)makeTouchBar
{
    NSTouchBar *bar = [[NSTouchBar alloc] init];
    bar.delegate = self;
    
    bar.customizationIdentifier = CandidateCustomizationIdentifier;
    
    // Set the default ordering of items.
    bar.defaultItemIdentifiers = @[NSTouchBarItemIdentifierCandidateList, NSTouchBarItemIdentifierOtherItemsProxy];
    
    bar.customizationAllowedItemIdentifiers = @[NSTouchBarItemIdentifierCandidateList];
    
    bar.principalItemIdentifier = NSTouchBarItemIdentifierCandidateList;
    
    return bar;
}

- (NSString *)nameFromContact:(CNContact *)contact
{
    NSMutableString *nameStr = [NSMutableString string];
    if (contact.givenName != nil && contact.givenName.length > 0)
    {
        nameStr = [NSMutableString stringWithFormat:@"%@ ", contact.givenName];
    }
    if (contact.familyName != nil && contact.familyName.length > 0)
    {
        [nameStr appendString:contact.familyName];
    }
    return nameStr;
}

- (NSString *)emailFromContact:(CNContact *)contact
{
    NSString *emailStr = @"";
    if (contact.emailAddresses != nil && contact.emailAddresses.count > 0)
    {
        CNLabeledValue *emailValue = contact.emailAddresses[0];
        emailStr = emailValue.value;
    }
    return emailStr;
}

#pragma mark NSTouchBarDelegate

// The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
- (nullable NSTouchBarItem *)touchBar:(NSTouchBar *)touchBar makeItemForIdentifier:(NSTouchBarItemIdentifier)identifier
{
    if ([identifier isEqualToString:NSTouchBarItemIdentifierCandidateList])
    {
        _candidateListItem = [[NSCandidateListTouchBarItem alloc] initWithIdentifier:NSTouchBarItemIdentifierCandidateList];
        self.candidateListItem.delegate = self;
        
        self.candidateListItem.customizationLabel = NSLocalizedString(@"Candidate List", @"");
        
        // Determine the appropriate descriptive string for this candidate.
        NSAttributedString *(^contactsCandidateStr)(CNContact *candidateContact, NSInteger index) = ^(CNContact *candidateContact, NSInteger index) {
            if (candidateContact != nil)
            {
                NSString *nameStr = [self nameFromContact:candidateContact];
                NSString *emailStr = [self emailFromContact:candidateContact];
                
                return [[NSAttributedString alloc] initWithString:[NSString stringWithFormat:@"%@\r<%@>", nameStr, emailStr]];
            }
            else
            {
                return [[NSAttributedString alloc] initWithString:@""];
            }
            
        };
        self.candidateListItem.attributedStringForCandidate = contactsCandidateStr;
        
        return self.candidateListItem;
    }
    return nil;
}


#pragma mark - NSControlSubclassNotifications

- (void)searchForCandidatesWithString:(NSString *)string
{
    NSError *error = nil;
    CNContactStore *addressBook = [[CNContactStore alloc] init];
    NSPredicate *searchPredicate = [CNContact predicateForContactsMatchingName:string];
    
    NSArray<CNContact *> *foundContacts = [addressBook unifiedContactsMatchingPredicate:searchPredicate keysToFetch:@[CNContactGivenNameKey, CNContactFamilyNameKey, CNContactEmailAddressesKey] error:&error];
    
    NSText *text = [self.view.window fieldEditor:NO forObject:self.textField];
    NSRange range = text.selectedRange;
    [self.candidateListItem setCandidates:foundContacts forSelectedRange:range inString:nil];
}

- (void)controlTextDidChange:(NSNotification *)notification
{
    NSTextField *textField = (NSTextField *)notification.object;
    
    if (textField.stringValue.length > 0)
    {
        CNEntityType entityType = CNEntityTypeContacts;
        if ([CNContactStore authorizationStatusForEntityType:entityType] == CNAuthorizationStatusNotDetermined)
        {
            CNContactStore *contactStore = [[CNContactStore alloc] init];
            [contactStore requestAccessForEntityType:entityType completionHandler:^(BOOL granted, NSError * _Nullable error) {
                if (granted)
                {
                    dispatch_async(dispatch_get_main_queue(), ^{
                        [self searchForCandidatesWithString:textField.stringValue];
                    });
                }
            }];
        }
        else if ([CNContactStore authorizationStatusForEntityType:entityType] == CNAuthorizationStatusAuthorized)
        {
            [self searchForCandidatesWithString:textField.stringValue];
        }
        else if ([CNContactStore authorizationStatusForEntityType:entityType] == CNAuthorizationStatusDenied)
        {
            static dispatch_once_t onceToken;
            dispatch_once(&onceToken, ^{
                // Run this alert only once.
                NSAlert *alert = [[NSAlert alloc] init];
                alert.messageText = NSLocalizedString(@"Access to Contacts is not allowed", comment:@"");
                alert.informativeText =
                    NSLocalizedString(@"Unable to search Contacts. Check System Preferences, \"Security & Privacy\" and give this application access.", comment:@"");
                [alert runModal];
            });
        }
    }
    else
    {
        [self.candidateListItem setCandidates:@[] forSelectedRange:NSMakeRange(0, 0) inString:nil];
    }
}


#pragma mark - NSCandidateListTouchBarItemDelegate

// The system invokes this when -candidateListVisible changes the visibility state.
- (void)candidateListTouchBarItem:(NSCandidateListTouchBarItem *)anItem changedCandidateListVisibility:(BOOL)isVisible
{
    //..
}

- (void)candidateListTouchBarItem:(NSCandidateListTouchBarItem *)anItem beginSelectingCandidateAtIndex:(NSInteger)index
{
    //..
}

// The system invokes this when the user moves from touching one candidate in the bar to another.
- (void)candidateListTouchBarItem:(NSCandidateListTouchBarItem *)anItem changeSelectionFromCandidateAtIndex:(NSInteger)previousIndex toIndex:(NSInteger)index
{
    //..
}

// The system invokes this when the user stops touching candidates in the bar. If index==NSNotFound, the user didn't select any candidate.
- (void)candidateListTouchBarItem:(NSCandidateListTouchBarItem *)anItem endSelectingCandidateAtIndex:(NSInteger)index
{
    if (anItem == self.candidateListItem)
    {
        CNContact *candidateContact = self.candidateListItem.candidates[index];
        NSString *nameStr = [self nameFromContact:candidateContact];
        NSString *emailStr = [self emailFromContact:candidateContact];
        self.textField.stringValue = [[NSString alloc] initWithString:[NSString stringWithFormat:@"%@ <%@>", nameStr, emailStr]];
    }
}

@end

File: CreatingAndCustomizingTheTouchBar/Swift/NSTouchBar Catalog/CandidateListViewController.swift
/*
See LICENSE folder for this sample’s licensing information.

Abstract:
View controller responsible for showing NSCandidateListTouchBarItem in an NSTouchBar instance for picking contacts as a use case.
*/

import Cocoa
import Contacts

class CandidateListViewController: NSViewController, NSTextFieldDelegate {
    
    let candidateListBar = NSTouchBar.CustomizationIdentifier("com.TouchBarCatalog.candidateListBar")
    var candidateListItem: NSCandidateListTouchBarItem<CNContact>!
    
    @IBOutlet weak var textField: NSTextField!
    
    // MARK: View Controller Life Cycle
    
    override func viewDidAppear() {
        super.viewDidAppear()
        
        // Use the candidate item instead of the NSTextField item.
        textField.isAutomaticTextCompletionEnabled = false
    }
    
    // MARK: NSTouchBarProvider
    
    override func makeTouchBar() -> NSTouchBar? {
        let touchBar = NSTouchBar()
        touchBar.delegate = self
        touchBar.customizationIdentifier = candidateListBar
        touchBar.defaultItemIdentifiers = [.candidateList, .otherItemsProxy]
        touchBar.customizationAllowedItemIdentifiers = [.candidateList]
        
        return touchBar
    }
    
    func candidateName(from contact: CNContact) -> String {
        if !contact.givenName.isEmpty {
            if !contact.familyName.isEmpty {
                return contact.givenName + " " + contact.familyName
            } else {
                return contact.givenName
            }
        } else {
            return contact.familyName
        }
    }
    
    func candidateEmail(from contact: CNContact) -> String {
        if !contact.emailAddresses.isEmpty {
            return contact.emailAddresses[0].value as String
        } else {
            return ""
        }
    }
    
    func candidates(matching name: String) -> [CNContact] {
        let keysToFetch = [CNContactGivenNameKey as CNKeyDescriptor,
                           CNContactFamilyNameKey as CNKeyDescriptor,
                           CNContactEmailAddressesKey as CNKeyDescriptor]
        do {
            let contacts =
                try CNContactStore().unifiedContacts(matching: CNContact.predicateForContacts(matchingName: name),
                                                     keysToFetch: keysToFetch)
            return contacts
        } catch {
            print("Caught an error but not handle it at \(#function): \(error)")
        }
        
        return []
    }
    
    private lazy var runContactsPermissionAlert: Void = {
        // Run this alert only once.
        let alert = NSAlert()
        alert.messageText = NSLocalizedString("Access to Contacts is not allowed", comment: "")
        alert.informativeText =
            NSLocalizedString("Unable to search Contacts. Check System Preferences, \"Security & Privacy\" and give this application access.",
                              comment: "")
        alert.runModal()
    }()
    
    func controlTextDidChange(_ notification: Notification) {
        guard let textField = notification.object as? NSTextField else { return }
        
        var candidates: [CNContact] = []
        
        let range: NSRange
        if let text = view.window?.fieldEditor(false, for: textField) {
            range = text.selectedRange
        } else {
            range = NSRange(location: 0, length: 0)
        }
        
        if !textField.stringValue.isEmpty {
            let authorizeStatus = CNContactStore.authorizationStatus(for: .contacts)
            
            if authorizeStatus == .notDetermined {
                CNContactStore().requestAccess(for: .contacts) { granted, _ in
                    if granted {
                        DispatchQueue.main.async {
                            candidates = self.candidates(matching: textField.stringValue)
                        }
                    }
                }
            } else if authorizeStatus == .denied {
                _ = runContactsPermissionAlert
            } else if authorizeStatus == .authorized {
                candidates = self.candidates(matching: textField.stringValue)
            }
        }
        
        candidateListItem.setCandidates(candidates, forSelectedRange: range, in: nil)
    }
}

// MARK: - NSTouchBarDelegate

extension CandidateListViewController: NSTouchBarDelegate {
    
    // The system calls this while constructing the NSTouchBar for each NSTouchBarItem you want to create.
    func touchBar(_ touchBar: NSTouchBar, makeItemForIdentifier identifier: NSTouchBarItem.Identifier) -> NSTouchBarItem? {
        guard identifier == .candidateList else { return nil }
        
        candidateListItem = NSCandidateListTouchBarItem<CNContact>(identifier: identifier)
        candidateListItem.delegate = self
        candidateListItem.customizationLabel = NSLocalizedString("Candidate List", comment: "")
        
        candidateListItem.attributedStringForCandidate = { [unowned self] (candidate, index) -> NSAttributedString in
            return NSAttributedString(string: self.candidateName(from: candidate) + "\r" + self.candidateEmail(from: candidate))
        }
        
        return candidateListItem
    }
}

// MARK: - NSCandidateListTouchBarItemDelegate

extension CandidateListViewController: NSCandidateListTouchBarItemDelegate {
    func candidateListTouchBarItem(_ anItem: NSCandidateListTouchBarItem<AnyObject>, endSelectingCandidateAt index: Int) {
        guard anItem == candidateListItem && anItem.candidates.count > index else { return }
        
        let contact = candidateListItem.candidates[index]
        textField.stringValue = candidateName(from: contact) + " <" + candidateEmail(from: contact) + ">"
    }
}

